File: ReverseLinkedList.java
class Node {
    int data;
    Node next;

    Node(int data, Node next) {
        this.data = data;
        this.next = next;
    }
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}
public class ReverseLinkedList {
    public static Node reverseLinkedList(Node head) {
        if (head == null || head.next == null) {
            return head;
        }
        Node newHead = reverseLinkedList(head.next);
        head.next.next = head;
        head.next = null;
        return newHead;
    }

    public static void printLinkedList(Node head) {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(3);
        head.next.next = new Node(2);
        head.next.next.next = new Node(4);
        
        System.out.println("The original linked list is :");
        printLinkedList(head);

        System.out.println("the reversed Linked list is ");
        head = reverseLinkedList(head);

        printLinkedList(head);
    }
}

File: reverseLinkedList.py
class Node:
    def __init__(self, data, next_node = None) -> None:
        self.data = data
        self.next = next_node

def reverseLinkedList(head):
    if head is None or head.next is None:
        return head
    newHead = reverseLinkedList(head.next)
    head.next.next = head
    head.next = None
    return newHead

def printLinkedList(head):
    temp = head
    while temp is not None:
        print(temp.data, end=" ")
        temp = temp.next
    print()
    

if __name__ == "__main__":
    head = Node(1)
    head.next = Node(3)
    head.next.next = Node(2)
    head.next.next.next = Node(4)
    
    print("Original Linked list is: ", end=" ")
    printLinkedList(head)
    
    print("Reversed Linked list is: ", end=" ")
    head = reverseLinkedList(head)
    
    printLinkedList(head)